File: robot-factory/python/tests/robot_order_builder.py
from __future__ import annotations

from robot_factory import PartOption, PartType, RobotOrder


class RobotOrderBuilder:
    def __init__(self) -> None:
        self._head: PartOption = PartOption.STANDARD_VISION
        self._body: PartOption = PartOption.SQUARE
        self._arms: PartOption = PartOption.HANDS
        self._movement: PartOption = PartOption.WHEELS
        self._power: PartOption = PartOption.SOLAR
        self._excluded: set[PartType] = set()

    def with_head(self, option: PartOption) -> RobotOrderBuilder:
        self._head = option
        return self

    def with_body(self, option: PartOption) -> RobotOrderBuilder:
        self._body = option
        return self

    def with_arms(self, option: PartOption) -> RobotOrderBuilder:
        self._arms = option
        return self

    def with_movement(self, option: PartOption) -> RobotOrderBuilder:
        self._movement = option
        return self

    def with_power(self, option: PartOption) -> RobotOrderBuilder:
        self._power = option
        return self

    def without(self, part_type: PartType) -> RobotOrderBuilder:
        self._excluded.add(part_type)
        return self

    def build(self) -> RobotOrder:
        order = RobotOrder()
        if PartType.HEAD not in self._excluded:
            order.configure(PartType.HEAD, self._head)
        if PartType.BODY not in self._excluded:
            order.configure(PartType.BODY, self._body)
        if PartType.ARMS not in self._excluded:
            order.configure(PartType.ARMS, self._arms)
        if PartType.MOVEMENT not in self._excluded:
            order.configure(PartType.MOVEMENT, self._movement)
        if PartType.POWER not in self._excluded:
            order.configure(PartType.POWER, self._power)
        return order

File: robot-factory/typescript/tests/RobotOrderBuilder.ts
import type { PartOption } from '../src/PartOption.js';
import type { PartType } from '../src/PartType.js';
import { RobotOrder } from '../src/RobotOrder.js';

export class RobotOrderBuilder {
  private head: PartOption = 'StandardVision';
  private body: PartOption = 'Square';
  private arms: PartOption = 'Hands';
  private movement: PartOption = 'Wheels';
  private power: PartOption = 'Solar';
  private readonly excluded = new Set<PartType>();

  withHead(option: PartOption): this { this.head = option; return this; }
  withBody(option: PartOption): this { this.body = option; return this; }
  withArms(option: PartOption): this { this.arms = option; return this; }
  withMovement(option: PartOption): this { this.movement = option; return this; }
  withPower(option: PartOption): this { this.power = option; return this; }
  without(type: PartType): this { this.excluded.add(type); return this; }

  build(): RobotOrder {
    const order = new RobotOrder();
    if (!this.excluded.has('Head')) order.configure('Head', this.head);
    if (!this.excluded.has('Body')) order.configure('Body', this.body);
    if (!this.excluded.has('Arms')) order.configure('Arms', this.arms);
    if (!this.excluded.has('Movement')) order.configure('Movement', this.movement);
    if (!this.excluded.has('Power')) order.configure('Power', this.power);
    return order;
  }
}
